apps/ejabberd/src/mod_push_mnesia.erl

```diff
@@ -0,0 +1,153 @@
+    ok.
+
+
+-spec enable(UserJID :: ejabberd:jid(), PubsubJID :: ejabberd:jid(),
+             Node :: mod_push:pubsub_node(), Form :: mod_push:form()) ->
+                    ok | {error, Reason :: term()}.
+enable(User, PubSub, Node, Forms) ->
+    disable(User, PubSub, Node),
+    write(make_record(User, PubSub, Node, Forms)).
+
+
+-spec disable(UserJID :: ejabberd:jid(), PubsubJID :: ejabberd:jid(),
+              Node :: mod_push:pubsub_node()) -> ok | {error, Reason :: term()}.
+disable(User, undefined, undefined) ->
+    delete(key(User));
+disable(User, PubsubJID, Node) ->
+    Result =
+        transaction(
+          fun() ->
+                  PubsubFiltered =
+                      [Record ||
+                          #push_subscription{pubsub_jid = RecPubsubJID} = Record <- read(key(User)),
+                          PubsubJID == undefined orelse RecPubsubJID == PubsubJID],
+
+                  NodeFiltered =
+                      [Record ||
+                          #push_subscription{pubsub_node = RecNode} = Record <- PubsubFiltered,
+                          Node == undefined orelse RecNode == Node],
+
+                  [mnesia:delete_object(Record) || Record <- NodeFiltered]
+          end),
+
+    case Result of
+        {error, _} = E ->
+            E;
+        _ ->
+            ok
+    end.
+
+
+-spec get_publish_services(User :: ejabberd:jid()) ->
+                                  {ok, [{PubSub :: ejabberd:jid(), Node :: mod_push:node(),
+                                         Form :: mod_push:form()}]} |
+                                  {error, Reason :: term()}.
+get_publish_services(User) ->
+    case safe_read(key(User)) of
+        {ok, Records} ->
+            {ok, [{PubsubJID, Node, Forms} ||
+                #push_subscription{pubsub_jid = PubsubJID,
+                                   pubsub_node = Node,
+                                   form = Forms} <- Records]};
+        {error, _} = E ->
+            E
+    end.
+
+%%--------------------------------------------------------------------
+%% Helper functions
+%%--------------------------------------------------------------------
+
+-spec read(key()) -> [sub_record()] | no_return().
+read(Key) ->
+    F = fun() -> mnesia:read({push_subscription, Key}) end,
+    mnesia:async_dirty(F).
+
+-spec safe_read(key()) -> {ok, [sub_record()]} | {error, Reason :: term()}.
+safe_read(Key) ->
+    try read(Key) of
+        Records ->
+            {ok, Records}
+    catch
+        _:Reason ->
+            {error, Reason}
+    end.
+
+-spec write(sub_record()) -> ok | {error, Reason :: term()}.
+write(Record) ->
+    F = fun() -> mnesia:write(Record) end,
+    transaction(F).
+
+-spec delete(key()) -> ok | {error, Reason :: term()}.
+delete(Key) ->
+    F = fun() -> mnesia:delete({push_subscription, Key}) end,
+    transaction(F).
+
+-spec transaction(fun(() -> any())) -> Result :: any().
+transaction(F) ->
```

Comment:
Do we need transactions in case of `mod_push`? I believe that conflicts or race conditions won't occur without them and it's always less work for Mnesia. :)